keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1154 @@
+[job tracking with finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/).
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+The `DeleteOptions` is extended by the reason field to convey the reason of
+deletion.
+
+```golang
+// DeleteOptions may be provided when deleting an API object.
+type DeleteOptions struct {
+  ...
+	// A brief CamelCase message indicating the reason why the object is deleted (example: "Eviction").
+	// Particularly useful when determining if a pod should be restarted after termination.
+	// + optional
+	Reason *string `json:"reason,omitempty" protobuf:"bytes,6,rep,name=reason"`
+}
+```
+
+The received `reason` will be stored in the already existing `status.reason` field in `PodStatus`:
+
+```
+// PodStatus represents information about the status of a pod. Status may trail the actual
+// state of a system.
+type PodStatus struct {
+  ...
+	// A brief CamelCase message indicating details about why the pod is in this state. e.g. 'Evicted'
+	// +optional
+	Reason string
+  ...
+}
+```
+
+### JobSpec API
+
+We extend the Job API in order to allow to apply different actions depending
+on the conditions associated with the pod failure.
+
+```golang
+// BackoffPolicyAction specifies how a Pod failure is handled
+type BackoffPolicyAction string
+
+const (
+
+	// This is an action which might be taken on pod failure - mark the
+	// pod's job as Failed
+	TerminateBackoffAction BackoffPolicyAction = "Terminate"
+
+	// This is an action which might be taken on pod failure - the pod will be
+	// restarted and the counter for .backoffLimit will not be incremented.
+	IgnoreBackoffAction BackoffPolicyAction = "Ignore"
+
+	// This is the default action which might be taken on pod failure -
+	// the counter for the backoffLimit is incremented and the pod is restarted
+	// if the backoffLimit is not reached yet. Once reached, the pod's job is
+	// marked as Failed
+	DefaultBackoffAction BackoffPolicyAction = "Default"
+)
+
+// BackoffPolicyActionWithConditions specifies the action taken on a pod failure
+// and the conditions which need to be satisfied to take the specified action
+type BackoffPolicyActionWithConditions struct {
+
+	// Specifies the action taken on a pod failure if all the specified
+	// conditions are satisfied
+	Action BackoffPolicyAction
+
+	// Restricts the check for exit codes to only apply to the container with
+	// the specified name.
+	// +optional
+	ContainerName *string
+
+	// Specifies the comma-separated list of exit codes or ranges of exit codes
+	// +optional
+	ContainerExitCode *string
+
+	// Specifies the comma-separated list of reasons for the pod failure
```

Comment:
It would be useful for compute providers to be able to say something like: "any infrastructure error" without having to list all of them. Most likely, the way to express this would be something like: anything that is not an empty reason, because when a pod fails on its own kubelet doesn't introduce a reason.

Alternatives are to use regex or an expression matcher (similar to a label selector) as we discussed offline.